### PR TITLE
Refactor and improve MLL tests

### PR DESCRIPTION
**Pull request by vitek-karas**
* Remove duplicated code - I know that tests should be descriptive, but repeating 100 times that we want to capture output doesn't feel necessary.
* For some of them move more stuff into the shared test state - this improves perf as we avoid repeating the setup (copying files around) for each test case, we do it once for the entire class
I also changed some of the tests to "Theory" as it's easier to read that way.
  * `SDKLookup.cs` - moved most of the state in to the shared state to speed up the tests.

* Adding more cases into the "theory" tests
  * Most notably for the framework resolution I added variations on the TFM (which will be needed when we implement disable of MLL)
* Adding new tests mostly around "list runtimes" (various forms), "list sdks" (various forms) and errors (which also list runtimes or sdks)

* Ported all of the `MultiLevelLookupFramework` tests over to the `FrameworkResolution` and `DependencyResolutions` suites which have a more robust test infra and can run the same tests much faster. Along the way I added lot more variations on top of the existing tests:
  * `PerAssemblyVersionResolutionMultipleFrameworks.cs` - this is actually not an MLL test, but I moved it to the new infra to make it much faster
  * `MultipleHives.cs` - MLL framework resolution tests

For SDK resolution I kept the `MultiLevelSDKLookup.cs` just removed code duplication and added new variants.

For the core reviewers: I promise I didn't remove any single test case in spirit with these exceptions:

* We had tests which validated that framework resolution is not affected by frameworks in current directory and also by frameworks in the user's directory. I left some basic test for the current directory check, but I completely removed the user's directory variant as the product simply doesn't have any code around that anymore.

### CODE REVIEW

**Comment by elinor-fung** [comment]
> completely removed the user's directory variant as the product simply doesn't have any code around that anymore

Could we do this for the SDK lookup tests too? I don't recall ever seeing anything around user directory for SDK resolution either.

**Comment by vitek-karas** [comment]
Good point - I removed the "user directory" parts from SDK resolution tests as well.

File: src/installer/tests/HostActivation.Tests/DependencyResolution/PerAssemblyVersionResolutionMultipleFrameworks.cs [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+        [InlineData(TestAssemblyWithFileVersion, null, "4.0.0.0", null)] // App has higher version than either framework
+        [InlineData(TestAssemblyWithFileVersion, null, "3.2.2.2", MicrosoftNETCoreApp)] // Exactly equal - lower one wins
+        public void AppWithSameAssemblyAsFramework(string testAssemblyName, string appAsmVersion, string appFileVersion, string frameWorkWins)
+        {
+            RunTest(null, testAssemblyName, appAsmVersion, appFileVersion, frameWorkWins);
+        }
+
+        [Theory]
+        [InlineData("1.1.1")]  // Exact match - no roll forward
+        [InlineData("1.1.0")]  // Path roll forward
```
Comment (elinor-fung) [inline src/installer/tests/HostActivation.Tests/DependencyResolution/PerAssemblyVersionResolutionMultipleFrameworks.cs:67]:
```suggestion
        [InlineData("1.1.0")]  // Patch roll forward
```

File: src/installer/tests/HostActivation.Tests/FrameworkResolution/MultipleHives.cs [hunk 2]
```diff
@@ -20,56 +22,211 @@ public MultipleHives(SharedTestState sharedState)
...
+
+            string expectedOutput = String.Join(
+                String.Empty,
+                GetExpectedFrameworks(multiLevelLookup)
+                    .Select(t => $"{MicrosoftNETCoreApp} {t.Version} [{Path.Combine(t.Path, "shared", MicrosoftNETCoreApp)}]{Environment.NewLine}"));
+
+            RunTest(new TestSettings()
+                    .WithCommandLine("--list-runtimes"),
+                multiLevelLookup,
+                testApp: null)
```
Comment (elinor-fung) [inline src/installer/tests/HostActivation.Tests/FrameworkResolution/MultipleHives.cs:157]: nit: indentation

File: src/installer/tests/HostActivation.Tests/DependencyResolution/PerAssemblyVersionResolutionMultipleFrameworks.cs [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+        [InlineData(TestAssemblyWithAssemblyVersion, "1.0.0.0", null, HighWare)]
+        [InlineData(TestAssemblyWithAssemblyVersion, null, "1.0.0.0", HighWare)]
+        [InlineData(TestAssemblyWithAssemblyVersion, "3.0.0.0", "1.0.0.0", null)] // App has higher version than any framework
+        [InlineData(TestAssemblyWithAssemblyVersion, "2.1.1.2", null, HighWare)] // Both are exactly the same, so lower level wins
+        [InlineData(TestAssemblyWithAssemblyVersion, "2.1.1.2", "1.0.0.0", null)]
+        [InlineData(TestAssemblyWithFileVersion, null, null, MicrosoftNETCoreApp)] // Frameworks both have the same version - lower one wins
+        [InlineData(TestAssemblyWithFileVersion, "1.0.0.0", null, null)] // App has assembly version, no framework has it - so app wins
+        [InlineData(TestAssemblyWithFileVersion, null, "1.0.0.0", MicrosoftNETCoreApp)]
+        [InlineData(TestAssemblyWithFileVersion, null, "4.0.0.0", null)] // App has higher version than either framework
+        [InlineData(TestAssemblyWithFileVersion, null, "3.2.2.2", MicrosoftNETCoreApp)] // Exactly equal - lower one wins
```
Comment (elinor-fung) [inline src/installer/tests/HostActivation.Tests/DependencyResolution/PerAssemblyVersionResolutionMultipleFrameworks.cs:59]: I feel like I've asked this before, but what does `lower` mean here?

Comment (elinor-fung) [reply src/installer/tests/HostActivation.Tests/DependencyResolution/PerAssemblyVersionResolutionMultipleFrameworks.cs:59]: Oh, this is the dependency hierarchy / depth thing, right? So app is highest level and NETCoreApp would always be lowest?

**Review by elinor-fung — APPROVED** [review]

File: src/installer/tests/HostActivation.Tests/DependencyResolution/PerAssemblyVersionResolutionMultipleFrameworks.cs (hunk 3 above)
Comment (vitek-karas) [reply src/installer/tests/HostActivation.Tests/DependencyResolution/PerAssemblyVersionResolutionMultipleFrameworks.cs:59]: Yes - the app is the highest level, NETCoreApp is always the "root" lowest level. In the resolutio logic, if everything else is equal the lowest level typically wins.

File: src/installer/tests/HostActivation.Tests/MultilevelSDKLookup.cs [hunk 4]
```diff
@@ -679,6 +467,103 @@ public void SdkMultilevelLookup_Must_Pick_The_Highest_Semantic_Version()
...
+            List<(string version, string rootPath)> expectedList = new();
+            expectedList.Add(("5.0.2", _exeSdkBaseDir));
+            expectedList.Add(("6.1.1", _exeSdkBaseDir));
+            expectedList.Add(("7.1.2", _exeSdkBaseDir));
+            if (multiLevelLookup is null || multiLevelLookup == true)
+            {
+                expectedList.Add(("6.2.0", _regSdkBaseDir));
+                expectedList.Add(("7.0.1", _regSdkBaseDir));
+            }
+            expectedList.Sort((a, b) => a.version.CompareTo(b.version));
```
[the remark this answers is not in the document]
Comment (vitek-karas) [reply src/installer/tests/HostActivation.Tests/MultilevelSDKLookup.cs:488]: Good point - fixed.

File: src/installer/tests/HostActivation.Tests/MultilevelSDKLookup.cs [hunk 5]
```diff
@@ -679,6 +467,103 @@ public void SdkMultilevelLookup_Must_Pick_The_Highest_Semantic_Version()
...
+        public void ListSdks(bool? multiLevelLookup)
+        {
+            // Multi-level lookup is only supported on Windows.
+            if (!OperatingSystem.IsWindows() && multiLevelLookup != false)
+                return;
+
+            var expectedList = AddSdkVersionsAndGetExpectedList(multiLevelLookup);
+            string expectedOutput = String.Join(String.Empty, expectedList.Select(t => $"{t.version} [{t.rootPath}]{Environment.NewLine}"));
+
+            RunTest("--list-sdks", multiLevelLookup)
```
[the remark this answers is not in the document]
Comment (vitek-karas) [reply src/installer/tests/HostActivation.Tests/MultilevelSDKLookup.cs:505]: Good idea - added.

**Review by elinor-fung — APPROVED** [review]

**Comment by vitek-karas** [comment]
The test failures are:
https://github.com/dotnet/runtime/issues/66174
https://github.com/dotnet/runtime/issues/65954

**Pull request merged**